### PR TITLE
postgresql: increase max length of display name before being truncated in username

### PR DESCRIPTION
**Pull request by beyondbill**
Fixes https://github.com/hashicorp/vault/issues/8560

Display name, including auth method `oidc` and full `sub` claim in our case, is at length of 25 characters. It will be truncated to the first 8 characters today in postgresql username. We won't be able to get the full `sub` claim out of postgresql username. Since `sub` claim is the only information in postgresql username that we can use to trace user identity, we need `sub` claim in its full form within postgresql username in order to run auditing if necessary.

Increasing the threshold of truncating to 30 will solve the issue. Beyond the context of `oidc` and `sub`, the benefit is the same in general: this PR will provide better information on who requested a postgresql user from Vault.

There's already a very similar [PR](https://github.com/hashicorp/vault/pull/8582) opened but it seems abandoned without testing. I have added unit tests to make sure username generation is fully covered.

### CODE REVIEW

**Comment by beyondbill** [comment]
Can someone take a look at this PR? It is currently blocking my project and I'd like to avoid using a fork if possible. Thanks!

**Comment by raskchanky** [comment]
@beyondbill I'll try to take a look tomorrow. Thank you for your patience!

File: sdk/database/helper/credsutil/sql_test.go
Comment (raskchanky) [inline sdk/database/helper/credsutil/sql_test.go:92]: If I'm reading [the code](https://github.com/hashicorp/vault/blob/master/sdk/database/helper/credsutil/sql.go#L41) correctly, `credsProducer.UsernameLen` is the maximum length for a username, so I think you can remove `expectedLen` and all of its associated logic and just replace this if statement with:
```go
if len(username) > credsProducer.UsernameLen {
```
and adjust the error message accordingly. I don't think it's necessarily important to know the exact username length for this test, and including too many internal details about the username generation algorithm will make the test more brittle.

**Review by raskchanky — CHANGES_REQUESTED** [review]
On the whole, I think this looks great! I left a few suggestions.

In addition, since this PR was opened back in July, it'd be awesome if you could rebase this onto the latest master branch. I wasn't able to actually run your tests until I did that.

**Comment by raskchanky** [comment]
@beyondbill In addition to my review comments, we'll need you to sign the CLA before this PR can be merged. Thanks!

File: sdk/database/helper/credsutil/sql_test.go
Comment (beyondbill) [reply sdk/database/helper/credsutil/sql_test.go:21]: @raskchanky Theoretically they don't necessarily have to match. But you are right. A different value in the test can be really confusing. I'll change it to match 🙂

File: sdk/database/helper/credsutil/sql_test.go
Comment (beyondbill) [reply sdk/database/helper/credsutil/sql_test.go:92]: @raskchanky Really good suggestion! I was going too far and detailed in testing username length.

**Comment by beyondbill** [comment]
All the tests I added have passed
```
PASS sdk/database/helper/credsutil.TestSQLGenerateUsername/both_display_name_and_role_name_are_empty (0.00s)
PASS sdk/database/helper/credsutil.TestSQLGenerateUsername/short_display_name_and_role_name_are_fully_shown (0.00s)
PASS sdk/database/helper/credsutil.TestSQLGenerateUsername/long_display_name_is_truncated (0.00s)
PASS sdk/database/helper/credsutil.TestSQLGenerateUsername/long_role_name_is_truncated (0.00s)
PASS sdk/database/helper/credsutil.TestSQLGenerateUsername/long_display_name_and_role_name_are_truncated (0.00s)
PASS sdk/database/helper/credsutil.TestSQLGenerateUsername/empty_display_name (0.00s)
PASS sdk/database/helper/credsutil.TestSQLGenerateUsername/empty_role_name (0.00s)
PASS sdk/database/helper/credsutil.TestSQLGenerateUsername (0.00s)
```
The two tests failed in `test-go-race` and should not have any relation with the change introduced in this PR.

**Comment by beyondbill** [comment]
@raskchanky Just saw https://github.com/hashicorp/vault/pull/10061 merged and significantly changed `plugins/database/postgresql/postgresql.go`... I'll try to find some time this evening to address it

**Comment by pcman312** [comment]
Hi @beyondbill! Thanks a lot for submitting this PR!

Unfortunately there are several problems with your proposed fix for #8560:

1. Changing the characteristics of how a username is generated will affect other users who are depending on the length or shape of the generated username. [Hyrum's Law](https://www.hyrumslaw.com/) unfortunately applies here:
    > With a sufficient number of users of an API, it does not matter what you promise in the contract: all observable behaviors of your system will be depended on by somebody.
2. Upping the display name limit to 30 characters will interfere with the maximum allowed length for similar reasons. Specifically, the username can have:
    - Display name: 30 characters
    - Role name: 8 characters
    - Random string: 20 characters
    - Timestamp: 10 characters
    - Prefix and separators: 5 characters

    This gives a maximum potential length of 73 characters which is then being truncated to 63 characters, removing the entire timestamp off of the username (and leaving a trailing separator). This will also be hit by Hyrum's Law. A regex matching the username is the obvious potential use case that makes this a problem.

3. This PR does not fix #8560. This only addresses the issue for PostgreSQL, not for all databases. We have been looking into the issue of customizing how usernames are generated and are currently planning on addressing this in a near-future release of Vault. 

Because of this, I am going to close this PR in favor of a more generic, long term solution for customizable username generation.

**Comment by beyondbill** [comment]
@pcman312 Thanks for the response! Although I'm sad to see the immediate solution that can unblock my current work gets rejected, I think you made good points and it's better to close this in favor of a long-term solution.

Can you tease more on the generic solution, e.g. PRs or docs? When you said a near-future release, is it targeted at Q4 2020 or a later time? I'll have to make a decision on a work project whether to wait for such a Vault release, use a forked Vault with this PR, or shelf the project for next year. Thank you!

**Comment by pcman312** [comment]
@beyondbill Thanks for understanding! I really hate closing this PR as it is useful (and plenty of others have asked for similar customization). We haven't started active work on the project yet (we've been gathering use cases and working on the current release), and unfortunately I can't give any specific timeframe as to when it will be completed. I'm currently envisioning giving the operator the ability to customize how a username is generated by using a Go template, however I haven't seen any specific set of requirements yet so this is very much subject to change.

**Comment by beyondbill** [comment]
@pcman312 Got it. I love the ability to customize how a username is generated by using a Go template!

**Comment by pcman312** [comment]
Hi @beyondbill ! I wanted to give an update on this. As we previously talked about, we have had a number of requests revolving around the idea of customizing how usernames are generated. In Vault 1.7 (coming soon) we are introducing the idea of `username_template` to some* of our secret engines. This will give the ability to fully customize how a username is generated prior to passing that username to the appropriate statements run in the database. Each engine will default to the current behavior, but this will allow for customization if that default does not suit their needs.

There will be docs available through https://www.vaultproject.io once 1.7 is released will describe this in more detail but I wanted to give you a heads up for this new feature coming soon!

<sub>* - engines supported in 1.7: PostgreSQL, MSSQL, MySQL, MongoDB (not MongoDB Atlas), Cassandra, Oracle, and OpenLDAP</sub>

**Pull request closed without merging**
